tsfc/driver.py

```diff
@@ -358,114 +348,14 @@ def compile_expression_dual_evaluation(expression, to_element, *,
-        dual_expressions = []   # one for each functional
-        broadcast_shape = len(expression.ufl_shape) - len(to_element.value_shape())
-        shape_indices = tuple(gem.Index() for _ in expression.ufl_shape[:broadcast_shape])
-        expr_cache = {}         # Sharing of evaluation of the expression at points
-        for dual in to_element.dual_basis():
-            pts = tuple(sorted(dual.get_point_dict().keys()))
-            try:
-                expr, point_set = expr_cache[pts]
-            except KeyError:
-                config = kernel_cfg.copy()
-                if runtime_quadrature_rule:
-                    # Until FInAT dual evaluation is done, FIAT
-                    # QuadratureElements with UnknownPointSet point sets
-                    # advertise NaNs as their points for each node in the dual
-                    # basis. This has to be manually replaced with the real
-                    # UnknownPointSet point set used to create the
-                    # QuadratureElement rule.
-                    assert isnan(pts).all()
-                    point_set = finat_to_element._rule.point_set
-                    config.update(point_indices=point_set.indices, point_expr=point_set.expression)
-                    context = fem.GemPointContext(**config)
-                else:
-                    point_set = PointSet(pts)
-                    config.update(point_set=point_set)
-                    context = fem.PointSetContext(**config)
-                expr, = fem.compile_ufl(expression, context, point_sum=False)
-                # In some cases point_set.indices may be dropped from expr, but
-                # nothing new should now appear
-                assert set(expr.free_indices) <= set(chain(point_set.indices, *argument_multiindices))
-                expr = gem.partial_indexed(expr, shape_indices)
-                expr_cache[pts] = expr, point_set
-            weights = collections.defaultdict(list)
-            for p in pts:
-                for (w, cmp) in dual.get_point_dict()[p]:
-                    weights[cmp].append(w)
-            qexprs = gem.Zero()
-            for cmp in sorted(weights):
-                qweights = gem.Literal(weights[cmp])
-                qexpr = gem.Indexed(expr, cmp)
-                qexpr = gem.index_sum(gem.Indexed(qweights, point_set.indices)*qexpr,
-                                      point_set.indices)
-                qexprs = gem.Sum(qexprs, qexpr)
-            assert qexprs.shape == ()
-            assert set(qexprs.free_indices) == set(chain(shape_indices, *argument_multiindices))
-            dual_expressions.append(qexprs)
-        basis_indices = (gem.Index(), )
-        ir = gem.Indexed(gem.ListTensor(dual_expressions), basis_indices)
+    # Get gem tensor with (num_nodes, ) shape.
+    ir_shape = to_element.dual_evaluation(fn)
+    broadcast_shape = len(expression.ufl_shape) - len(to_element.value_shape)
```

Comment:
delete this line - we deal with this in the `dual_evaluation` method of a tensor finite element